### ISSUE TITLE
Unexpected type error for accessing known existent members in a ReadonlyArray with noUncheckedIndexedAccess

### ISSUE THREAD

**Issue by chrisirhc** [microsoft/TypeScript#46922]
# Bug Report

### 🔎 Search Terms

non-existent members, ReadonlyArray
Referenced https://github.com/microsoft/TypeScript/issues/13778#issuecomment-950196264 .

### 🕗 Version & Regression Information

`noUncheckedIndexedAccess` was introduced in 4.1.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAHgLhgJQKYEMAm4A2BPAQQCcj08AeMAVwFsAjVIgPhgF4YBtABgBoYBGPgCYAugG4AUKEixgSavUZt43cUA)

### 💻 Code

```ts
const x: ReadonlyArray<number> = [0, 1, 2];
const c: number = x[0]; /* Unexpected `Type 'number | undefined' is not assignable to type 'number'.
  Type 'undefined' is not assignable to type 'number'.`
*/
```

### 🙁 Actual behavior

It returned `number | undefined` type for all members of the ReadonlyArray.

### 🙂 Expected behavior

Expected known members to be typed `number`.

**Comment by tadhgmister** [comment]
I think this is expected, `x[10]` is also `number | undefined` since your type doesn't retain length information.  `[0,1,2] as const` retains the length information and therefore `x[0]` is typed as number.

**Comment by fatcerberus** [comment]
That's right, `ReadOnlyArray` is just a generic array type and doesn't include the length of the array as part of the type, nor does the control flow analyzer track that information (hence why noUncheckedIndexedAccess is optional).  If you had gotten this array as an argument to a function call, for example, nothing in its typing says those elements have to be present, so the error is correct.

**Comment by RyanCavanaugh** [comment]
`ReadOnly` also doesn't mean immutable. You could get back an array from a function that still is able to mutate it later.

The behavior you expect can be achieved from
```ts
const x = [0, 1, 2] as const;
const c: number = x[0]; 
```
